### PR TITLE
Add internal documentation for the beam_makeops script

### PR DESCRIPTION
**Pull request by bjorng**
This PR adds internal documentation for the `beam_makeops` script, which is used when building the Erlang compiler and Erlang runtime system. I try to explain all about generic and specific instructions, transformation rules, and the new macro-based system for implementing BEAM instructions.

There are also a few commits with minor clean ups of issues found while writing the document.

### CODE REVIEW

File: erts/emulator/internal_doc/beam_makeops.md [hunk 1]
```diff
@@ -0,0 +1,1835 @@
...
+uppercase letter.  In the body, the macro arguments can be expanded
+by preceding them with `$`.
+
+A macro definition whose name and arity matches a family of
+specific instructions is assumed to be the implementation of that
+instruction.
+
+A macro can also be invoked from within another macro.  For example,
+`move_deallocate_return/2` avoids repeating code by invoking
+`$deallocate_return()` as a macro:
```
Comment (garazdawi) [inline erts/emulator/internal_doc/beam_makeops.md]: maybe mention that the arguments to macro calls cannot be complex expressions.

i.e.

    $deallocate_return(get_deallocation(y, $Deallocate));

will not work as the script splits arguments on `,` and does not care about () nesting.

File: erts/emulator/internal_doc/beam_makeops.md [hunk 2]
```diff
@@ -0,0 +1,1835 @@
...
+instruction.  We will show each individual fragment first, and then
+show how to connect them together.  First we will need a variable that
+we can store the value fetched from the register in:
+
+    increment.head() {
+        Eterm increment_reg_val;
+    }
+
+The name `increment` is the name of the group that the fragment
+belongs to.  Note that it does not need to have the same same
```
Comment (garazdawi) [inline erts/emulator/internal_doc/beam_makeops.md]: "same same"

**Comment by bjorng** [comment]
@garazdawi Thanks for the review. I have addressed your comments.

**Comment by bjorng** [comment]
I have added a commit that adds an additional test to the loading of `s` operands.

**Pull request merged**